packages/frontend-shared/src/utils/isAllowedFeature.ts

```diff
@@ -0,0 +1,104 @@
+  isLoggedOut: BannerIds.ACI_082022_LOGIN,
+  needsOrgConnect: BannerIds.ACI_082022_CREATE_ORG,
+  needsProjectConnect: BannerIds.ACI_082022_CONNECT_PROJECT,
+  needsRecordedRun: BannerIds.ACI_082022_RECORD,
+}
+
+/**
+ *
+ * @param eventTime - timestamp of an event
+ * @param waitTime - time to compare with, such as `1 day`, `20 minutes`, etc, to be parsed by `human-interval` package
+ */
+const minTimeSinceEvent = (eventTime: number | undefined, waitTime: string) => {
+  if (!eventTime) {
+    // if there is no event time, the event has never been recorded
+    // so the "time since" is infinite, return true here
+    return true
+  }
+
+  // converting to lowercase since `interval` will error on "Day" vs "day"
+  const waitTimestamp = interval(waitTime.toLocaleLowerCase())
+
+  if (isNaN(waitTimestamp)) {
+    throw new Error(`incorrect format for waitTime provided, value must be \`n days\`, \`n minutes\` etc. Value received was ${waitTime}`)
+  }
+
+  return (Date.now() - eventTime) > interval(waitTime)
+}
+
+export const isAllowedFeature = (
+  featureName: 'specsListBanner' | 'docsCiPrompt',
+  loginConnectStore: LoginConnectStore,
+) => {
+  const {
+    firstOpened,
+    promptsShown,
+    latestBannerShownTime,
+    bannersState,
+    userStatus,
+    hasNonExampleSpec,
+    userStatusIsNot,
+  } = loginConnectStore
+
+  const events = {
+    cypressFirstOpened: firstOpened,
+    navCiPromptAutoOpened: promptsShown.ci1,
+    loginModalRecordPromptShown: promptsShown.loginModalRecord,
+    latestSmartBannerShown: latestBannerShownTime,
+  }
+
+  function bannerForCurrentStatusWasNotDismissed () {
+    const bannerId = bannerIds[userStatus]
+
+    return !bannersState?.[bannerId]?.dismissed
+  }
+
+  function bannersAreNotDisabledForTesting () {
+    return !bannersState?._disabled
+  }
+
+  // For each feature, we define an array of rules for every `UserStatus`.
+  // The `base` rule is applied to all statuses, additional rules are
+  // nested in their respective statuses.
+  const rules = {
+    specsListBanner: {
+      base: [
+        minTimeSinceEvent(events.cypressFirstOpened, '4 days'),
+        minTimeSinceEvent(events.navCiPromptAutoOpened, '1 day'),
+        bannerForCurrentStatusWasNotDismissed(),
+        userStatusIsNot('allTasksCompleted'),
+        bannersAreNotDisabledForTesting(),
+      ],
+      needsRecordedRun: [
+        minTimeSinceEvent(events.loginModalRecordPromptShown, '1 day'),
+        hasNonExampleSpec,
+      ],
+      needsOrgConnect: [],
+      needsProjectConnect: [],
+      isLoggedOut: [],
+    },
+    docsCiPrompt: {
+      base: [
+        minTimeSinceEvent(events.cypressFirstOpened, '4 days'),
+        minTimeSinceEvent(events.latestSmartBannerShown, '1 day'),
+      ],
+      needsRecordedRun: [],
+      needsOrgConnect: [],
+      needsProjectConnect: [],
+      isLoggedOut: [],
+    },
+  }
+
+  let rulesToCheck = [...rules[featureName].base]
+
+  if (userStatus !== 'allTasksCompleted') {
```

Comment:
`base` is per feature, the docs CI prompt currently shows after 4 days regardless of the user's connection status. Which makes some sense, you can record a run locally, so "all tasks are completed", but we don't know if you have connected CI. So that's the where we fall through and only want to apply the base rules.

I think `allTasksCompleted` should be treated more like the rest, it's weird that I put it in the base rule. It kind of evolved that way. Will update.

Reply:
I restructured this a little and was able to remove the `userStatusIsNot` helper and this check. Let me know what you think of new version, I tried a couple ideas and settled on what I have now as a reasonable balance.